### DISCUSSION TITLE
Generic JSON to CSV Conversion, with Nested Arrays

### DISCUSSION THREAD

**Discussion by osevill** [TomWright/dasel#336] [General]
Would like to use dasel to convert from JSON to CSV, when the JSON file has nested arrays, and without specifying field names in the dasel expression, so that it will work on any JSON file with the same nested array format.

for example, starting with this array:
[
  {
    "string1": "abc",
    "string2": "123",
    "array1": [
      {
        "Key": "a",
        "Value": 1
      },
      {
        "Key": "b",
        "Value": 2
      }
    ]
  },
  {
    "string1": "def",
    "string2": "456",
    "array1": [
      {
        "Key": "c",
        "Value": 3
      },
      {
        "Key": "d",
        "Value": 4
      }
    ]
  }
]

... I end up with something like this (converted using jq):
"array1.0.Key","array1.0.Value","array1.1.Key","array1.1.Value","string1","string2"
"a",1,"b",2,"abc","123"
"c",3,"d",4,"def","456"

The above 2 elements have the same keys, but ideally dasel would collect the unique keys across all elements of the array when generating the CSV header row.  Either lexically sorting the unique keys, as in the example above, or building the header in found order of unique keys is fine, whichever is easier.

Thanks.

**Comment by KeelMatrix** [comment]
Short version: dasel's CSV writer can't produce the union header you're after, and the reason is worth knowing before you build a pipeline around it. In the writer (`parsing/csv/writer.go`) the header row is taken from the **first** record's keys, and every later record is then looked up by exactly those keys — so a key that only exists in record 2, or an `array1.1.*` that only exists when an array has two or more elements, can't be represented. A dasel expression rewrites one record at a time; collecting the keys across all records needs a pass over the whole array.

Here's the two-pass version in jq. Save as `flatten.jq`:

```jq
# JSON array (or concatenated objects) -> CSV with a union header.
# objects -> dotted columns (a.b), arrays -> positional columns (a.0.b)
def flatten($p):
  if   type == "object" then reduce (to_entries[]) as $e ({}; . + ($e.value | flatten($p + [$e.key])))
  elif type == "array"  then reduce (to_entries[]) as $e ({}; . + ($e.value | flatten($p + [($e.key|tostring)])))
  else { ($p | join(".")): . }
  end;

[ .[] | flatten([]) ] as $rows
| ( [ $rows[] | keys[] ] | unique ) as $cols      # union header (sorted)
| ( $cols | @csv ), ( $rows[] | [ .[$cols[]] ] | @csv )
```

```console
$ jq -r -f flatten.jq input.json
"array1.0.Key","array1.0.Value","array1.1.Key","array1.1.Value","string1","string2"
"a",1,"b",2,"abc","123"
"c",3,"d",4,"def","456"
```

If you'd rather not depend on jq, the same thing in Python (it also accepts concatenated objects, not just an array):

```python
import csv, json, sys

def flatten(obj, prefix=""):
    if isinstance(obj, dict):
        for k, v in obj.items():
            yield from flatten(v, f"{prefix}.{k}" if prefix else str(k))
    elif isinstance(obj, list):
        for i, v in enumerate(obj):
            yield from flatten(v, f"{prefix}.{i}" if prefix else str(i))
    else:
        yield prefix, obj

def records(stream):                      # JSON array *or* concatenated values
    data = stream.read()
    dec, i, n = json.JSONDecoder(), 0, len(data)
    while i < n:
        while i < n and data[i] in " \t\r\n,":
            i += 1
        if i >= n:
            return
        obj, i = dec.raw_decode(data, i)
        yield from (obj if isinstance(obj, list) else [obj])

rows = [dict(flatten(rec)) for rec in records(sys.stdin)]
cols = sorted({k for row in rows for k in row})
w = csv.writer(sys.stdout, lineterminator="\n")
w.writerow(cols)
for row in rows:
    w.writerow([row.get(c, "") for c in cols])
```

What to watch out for with this shape:

- **Ragged arrays.** Three elements in record 1 and one in record 2 gives `array1.0.*`, `array1.1.*`, `array1.2.*` with empty cells for the short record. Because `keys` is sorted, the header is at least stable across runs.
- **The wide shape is lossy in a specific way.** `array1.0.Key` and `array1.1.Key` no longer say "these two belong to the same parent" — the grouping is only implied by the column names. If the consumer needs real rows, the other shape is one row per array element with `string1`/`string2` repeated, but then the row count changes, so label the output for whoever reads it.
- **Empty arrays leave no trace** in the header, so a record containing only empty arrays still emits a row of empty cells.

Disclosure: I work on jsonnorm.com, an open JSON→CSV converter. Nested arrays are exactly the choice it makes explicit — keep them as indexed columns like above, or explode to one row per element with the row multiplication shown before download: https://jsonnorm.com/convert/nested-json-to-csv/?src=ghdisc
